### PR TITLE
#881: CreateRequestContainer rise system exception instead of ODataEx…

### LINKED ISSUE #881: CreateRequestContainer rise system exception instead of ODataException.
CreateRequestContainer rise system exception instead of ODataException 

### Assemblies affected
System.Web.OData.dll

### Reproduce steps
First of all it's strange that the following code can rise an exception in OData.      

> var context = new HttpContextWrapper(HttpContext.Current);
>       var routeData = RouteTable.Routes.GetRouteData(context);
>       var routeData2 = RouteTable.Routes.GetRouteData(context);

In new version of OData (6.0.0), inside of `HttpRequestMessageExtensions.CreateRequestContainer` throws InvalidOperationException (why we need the check?) if `request.Properties` already has requestContainer property. So, just call  `RouteTable.Routes.GetRouteData(context)` two times in a row, we have got such exception which not always possible to catch, because the call of CreateRequestContainer might be chained with a lot of different external libs. The `CreateRequestContainer` calls from `Match` method of `ODataPathRouteConstraint` class. The Match class contains try/catch logic, but it catch only ODataException. To make the Odata code more predictable, need to change type of exception, from InvalidException to OdataException (and still write Odata's message).
### Expected result
The 'Match' method from ODataPathRouteConstraint will catch OdataExceptions.

### Actual result
The 'Match' method doesn't catch exception throwed by `System.Web.OData.Extensions.HttpRequestMessageExtensions` class.

### Proposed fix:
Change `throw Error.InvalidOperation(SRResources.RequestContainerAlreadyExists);` 
to: `throw new ODataException(SRResources.RequestContainerAlreadyExists);`.

### PR DESCRIPTION
**Pull request by smitbmx**
…ception.

Done.

### Issues
#881

### Description
Fixed #881.

### CODE REVIEW

File: OData/src/System.Web.OData/Extensions/HttpRequestMessageExtensions.cs [hunk 1]
```diff
@@ -261,7 +261,7 @@ public static IServiceProvider CreateRequestContainer(this HttpRequestMessage re
         {
             if (request.Properties.ContainsKey(RequestContainerKey))
             {
-                throw Error.InvalidOperation(SRResources.RequestContainerAlreadyExists);
+                throw new ODataException(SRResources.RequestContainerAlreadyExists);
```
Comment (AlanWong-MS) [inline OData/src/System.Web.OData/Extensions/HttpRequestMessageExtensions.cs]: Please add an exception section in the function comments. For example,
```
        /// <exception cref="ODataException">
        /// <paramref name="request"/> already has requestContainer property.
        /// </exception>
```

**Comment by AlanWong-MS** [comment]
Thanks for your contribution! Before we can complete this request, can you please add tests to your PR?

File: OData/src/System.Web.OData/Extensions/HttpRequestMessageExtensions.cs (hunk 1 above)
Comment (smitbmx) [inline OData/src/System.Web.OData/Extensions/HttpRequestMessageExtensions.cs]: Done.

**Comment by biaol-odata** [comment]
@AlanWong-MS : It seems like the PR is still not merged yet? can you please tell us what is still required to be done? Thanks!

**Comment by AlanWong-MS** [comment]
@biaol-odata @smitbmx : we just need to update/add appropriate unit tests to accommodate this change.

**Comment by xuzhg** [comment]
Close it. We'd love to see your further PRs. Thanks your contribution.

**Pull request closed without merging**